### PR TITLE
Implement permute_dims

### PR DESCRIPTION
**Pull request by AllanZyne**
- very low efficient implementation in distributed mode (algorithm need to improve in the future)
- fix some build warnings
- support building & running sharpy using imex build directory directly (without deploy, occupy too many disk space)

### CODE REVIEW

File: src/idtr.cpp [hunk 1]
```diff
@@ -570,6 +575,385 @@ _idtr_copy_reshape(SHARPY::Transceiver *tc, int64_t iNSzs, void *iGShapeDescr,
       oData.data(), oData.sizes(), oData.strides());
 }
 
+namespace {
+
+class id {
```
Comment (fschlimb) [inline src/idtr.cpp:580]: Please add some high-level description

File: src/idtr.cpp [hunk 2]
```diff
@@ -570,6 +575,385 @@ _idtr_copy_reshape(SHARPY::Transceiver *tc, int64_t iNSzs, void *iGShapeDescr,
...
+
+  id operator-(const id &rhs) const {
+    std::vector<int64_t> new_values(_values.size());
+    for (size_t i = 0; i < _values.size(); i++) {
+      new_values[i] = _values[i] - rhs._values[i];
+    }
+    return id(std::move(new_values));
+  }
+
+  id operator-(const int64_t *rhs) const {
```
Comment (fschlimb) [inline src/idtr.cpp:601]: Some explanation would be good.

File: src/idtr.cpp [hunk 3]
```diff
@@ -570,6 +575,385 @@ _idtr_copy_reshape(SHARPY::Transceiver *tc, int64_t iNSzs, void *iGShapeDescr,
...
+    std::vector<int64_t> new_values(_values.size());
+    for (size_t i = 0; i < _values.size(); i++) {
+      new_values[i] = _values[i] - rhs[i];
+    }
+    return id(std::move(new_values));
+  }
+
+  void next(const int64_t *shape) {
```
Comment (fschlimb) [inline src/idtr.cpp:609]: Some explanation would be good.

File: src/idtr.cpp [hunk 4]
```diff
@@ -570,6 +575,385 @@ _idtr_copy_reshape(SHARPY::Transceiver *tc, int64_t iNSzs, void *iGShapeDescr,
...
+    for (size_t i = 0; i < _values.size(); i++) {
+      new_values[i] = _values[axes[i]];
+    }
+    return id(std::move(new_values));
+  }
+
+  int64_t operator[](size_t i) const { return _values[i]; }
+  int64_t &operator[](size_t i) { return _values[i]; }
+
+  id operator-(const id &rhs) const {
```
Comment (fschlimb) [inline src/idtr.cpp:593]: Some explanation would be good.

File: examples/transposed3d.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import numpy as np
```
Comment (fschlimb) [inline examples/transposed3d.py:1]: I think these should actually be tests. Could you move them to dir `test` and make them proper pytest tests?

**Review by fschlimb — CHANGES_REQUESTED** [review]

**Comment by fschlimb** [comment]
Very nice!

File: src/idtr.cpp (hunk 1 above)
Comment (AllanZyne) [reply src/idtr.cpp:578]: Added

File: src/idtr.cpp (hunk 2 above)
Comment (AllanZyne) [reply src/idtr.cpp:607]: Added

File: src/idtr.cpp (hunk 3 above)
Comment (AllanZyne) [reply src/idtr.cpp:619]: Added

File: src/idtr.cpp (hunk 4 above)
Comment (AllanZyne) [reply src/idtr.cpp:598]: Added

**Comment by AllanZyne** [comment]
@tkarna, I have updated the transpose.py, please check

File: sharpy/__init__.py [hunk 6]
```diff
@@ -96,6 +96,17 @@ def _validate_device(device):
         raise ValueError(f"Invalid device string: {device}")
 
 
+def arange(start, stop=None, step=1, dtype=int64, device="", team=1):
+    if stop is None:
+        stop = start
+        start = 0
```
Comment (tkarna) [inline sharpy/__init__.py:102]: Here you are adding a new way to call arange: `arange(stop)` which is equal to `arange(0, stop)`. This is useful, but this route should be tested too. Please add a test in test/test_create.py. Alternatively you can drop this change from this PR (unrelated to permute_dims).

**Review by tkarna — CHANGES_REQUESTED** [review]

File: sharpy/__init__.py (hunk 6 above)
Comment (AllanZyne) [reply sharpy/__init__.py]: removed

File: examples/transpose.py [hunk 7]
```diff
@@ -45,7 +45,8 @@ def sp_transpose(arr):
 
 
 def np_transpose(arr):
-    return numpy.ravel(arr.transpose()).reshape(arr.shape)
+    brr = arr.transpose()
+    return numpy.ravel(brr).reshape(brr.shape)
```
[the remark this answers is not in the document]
Comment (AllanZyne) [reply examples/transpose.py]: updated

**Review by tkarna — APPROVED** [review]
Looks good, thank you!

File: examples/transpose.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+
+def info(s):
+    if comm_rank == 0:
+        print(s)
+
+
+def sp_transpose(arr):
+    brr = sharpy.permute_dims(arr, [1, 0])
+    sharpy.sync()
```
Comment (fschlimb) [inline examples/transpose.py:43]: Why do we need a sync here?

**Review by fschlimb — DISMISSED** [review]
Yey!

**Review by tkarna — APPROVED** [review]

**Pull request merged**